posthog/tasks/alerts/checks.py

```diff
@@ -0,0 +1,108 @@
+from celery import shared_task
+from celery.canvas import group, chain
+from django.utils import timezone
+import math
+import structlog
+
+from posthog.api.services.query import ExecutionMode
+from posthog.caching.calculate_results import calculate_for_query_based_insight
+from posthog.email import EmailMessage
+from posthog.hogql_queries.legacy_compatibility.flagged_conversion_manager import (
+    conversion_to_query_based,
+)
+from posthog.models import Alert
+from posthog.schema import AnomalyCondition
+
+logger = structlog.get_logger(__name__)
+
+
+def check_all_alerts() -> None:
+    alert_ids = list(Alert.objects.all().values_list("id", flat=True))
+
+    group_count = 10
+    # All groups but the last one will have a group_size size.
+    # The last group will have at most group_size size.
+    group_size = int(math.ceil(len(alert_ids) / group_count))
+
+    groups = []
+    for i in range(0, len(alert_ids), group_size):
+        alert_id_group = alert_ids[i : i + group_size]
+        chained_calls = chain([check_alert_task.si(alert_id) for alert_id in alert_id_group])
+        groups.append(chained_calls)
+
+    group(groups).apply_async()
+
+
+def check_alert(alert_id: int) -> None:
+    alert = Alert.objects.get(pk=alert_id)
+    insight = alert.insight
+
+    with conversion_to_query_based(insight):
+        calculation_result = calculate_for_query_based_insight(
+            insight,
+            execution_mode=ExecutionMode.RECENT_CACHE_CALCULATE_BLOCKING_IF_STALE,
+            user=None,
+        )
+
+    if not calculation_result.result:
+        raise RuntimeError(f"No results for alert {alert.id}")
+
+    anomaly_condition = AnomalyCondition.model_validate(alert.anomaly_condition)
+    thresholds = anomaly_condition.absoluteThreshold
+
+    result = calculation_result.result[0]
+    aggregated_value = result["aggregated_value"]
+    anomalies_descriptions = []
+
+    if thresholds.lower is not None and aggregated_value < thresholds.lower:
+        anomalies_descriptions += [
+            f"The trend value ({aggregated_value}) is below the lower threshold ({thresholds.lower})"
+        ]
+    if thresholds.upper is not None and aggregated_value > thresholds.upper:
+        anomalies_descriptions += [
+            f"The trend value ({aggregated_value}) is above the upper threshold ({thresholds.upper})"
+        ]
+
+    if not anomalies_descriptions:
+        logger.info("No threshold met", alert_id=alert.id)
+        return
+
+    send_notifications(alert, anomalies_descriptions)
+
+
+@shared_task(ignore_result=True)
+def check_all_alerts_task() -> None:
+    check_all_alerts()
+
+
+@shared_task(ignore_result=True)
+def check_alert_task(alert_id: int) -> None:
+    check_alert(alert_id)
+
+
+# TODO: make it a task
```

Comment:
I don't think this needs to be a task. The `.send` function by default will queue a celery task for the actual sending.

Reply:
It makes sense, thanks, removed